crypto/src/main/java/io/warp10/crypto/OrderPreservingBase64.java

```diff
@@ -64,36 +75,54 @@ public static String encodeToString(byte[] data) {
    *
    * @param data   the data
    * @param offset the offset
-   * @param len    the len
+   * @param datalen    the len
    * @return the byte [ ]
    */
-  public static byte[] encode(byte[] data, int offset, int len) {
-    byte[] encoded = new byte[4 * (len / 3) + (len % 3 != 0 ? 1 + (len % 3) : 0)];
-
+  public static byte[] encode(byte[] data, int offset, int datalen) {
+    int i = 0;
+    int o = 0;
+    
+    int len = 4 * (datalen / 3) + (datalen % 3 != 0 ? 1 + (datalen % 3) : 0);
+    
+    byte[] encoded = new byte[len];
+    
     int idx = 0;
-
-    for (int i = 0; i < len; i++) {
+    
+    // first, process input per 3 bytes
+    int len24b = (datalen / 3) * 3; // length of 24 bits multiple
+    for (i = offset; i < (offset + len24b); i += 3) {
+      o = ((((data[i]) << 4) | ((data[i + 1] & 0xFF) >>> 4)) << 1) & 0x1FFF;
```

Comment:
Explain how the index is computed, the final left shift by 1 is only obvious if you know how the look up table is built, which so far is not described.

The computation could be simplified into"

(data[i] << 5 | ((data[i + 1] & 0xF0) >>> 3)) & 0x1FFF